File: leetcode 1625.py
import math
class Solution:
    def findLexSmallestString(self, s: str, a: int, b: int) -> str:
        s = list(map(int, s))
        n = len(s)
        g = math.gcd(a, 10)
        ans = [math.inf] 
        step = math.gcd(n, b)
        def modify(t, idx: int):
            num = t[idx]
            # target: num % g
            changed = (num % g - num + 10) % 10
            if changed:
                for j in range(idx, n, 2):
                    t[j] = (t[j] + changed) % 10
        for i in range(0, n, step):
            tmp = s[i:] + s[:i]
            modify(tmp, 1)
            if step % 2 != 0:
                modify(tmp, 0)
            if not ans or ans > tmp:
                ans = tmp
        return "".join(map(str, ans))

File: leetcode 1625.cpp
using PII = std::pair<int, int>;
using string = std::string;
class Solution {
public:
  string findLexSmallestString(string s, int a, int b) {
    int n = s.length();
    // step = k * gcd(n, b)
    string ans;
    int step = std::gcd(n, b);
    int g = std::gcd(10, a);
    auto modify = [&](string &tmp, int idx) -> void {
      int num = tmp[idx] - '0';
      int changed = (num % g - num) + 10;
      if (changed) {
        for (int i = idx; i < n; i += 2) {
          tmp[i] = (tmp[i] - '0' + changed) % 10 + '0';
        }
      }
    };
    for (int i = 0; i < n; i += step) {
      string tmp = s.substr(i) + s.substr(0, i);
      modify(tmp, 1);
      if (step & 1)
        modify(tmp, 0);
      if (ans.empty() || ans > tmp)
        ans = tmp;
    }
    return ans;
  }
};